perf(engine): deduplicate state_provider_builder calls

yongkangc opened the pull request:
Closes #20435

**Problem**
Two redundant `state_provider_builder` calls during block validation - one in `EngineApiTreeHandler::insert_block_inner` and another in `BasicEngineValidator::validate_block_with_state`. Each call traverses `blocks_by_hash` to build the overlay Vec.

**Solution**
Pass both the prebuilt `StateProviderBox` and the `StateProviderBuilder` through `TreeCtx` to avoid redundant lookups.

**Impact**
- Eliminates 1× `blocks_by_hash` HashMap traversal per block validation
- Eliminates 1× `Vec<ExecutedBlock>` allocation per block validation  

**Bench** ([code](https://gist.github.com/yongkangc/8159d84558a84f12550708140f237005))

| Overlay Size | single_lookup | double_lookup | Speedup |
|--------------|---------------|---------------|---------|
| 1 block | 10.18 ns | 20.15 ns | **2.0x** |
| 10 blocks | 102.82 ns | 203.91 ns | **2.0x** |
| 50 blocks | 510.36 ns | 1.04 µs | **2.0x** |
| 100 blocks | 1.24 µs | 2.59 µs | **2.1x** |

TL;DR: **~2x faster** for the builder step by avoiding the second lookup/build.

![state-provider-builder-lines](https://gist.githubusercontent.com/yongkangc/7f630dc08847b7737239ad11a71c7cfc/raw/lines.svg)
![state-provider-builder-violin](https://gist.githubusercontent.com/yongkangc/7f630dc08847b7737239ad11a71c7cfc/raw/violin.svg)

yongkangc committed:
perf(engine): deduplicate state_provider_builder calls

Avoid redundant state_provider_builder lookups during block validation by
passing both the prebuilt StateProviderBox and the builder through TreeCtx.

Key changes:
- StateProviderBuilder::build() now consumes self, avoiding Vec clone
- TreeCtx stores PrecomputedState with both provider and type-erased builder
- EngineApiTreeHandler builds provider immediately and passes both to validator
- Validator uses prebuilt provider for execution, builder for parallel tasks

This eliminates duplicate blocks_by_hash traversals that were occurring
during payload validation.

Closes #20435

mediocregopher commented on crates/engine/tree/src/tree/payload_validator.rs:
```diff
@@ -60,13 +60,38 @@ pub struct TreeCtx<'a, N: NodePrimitives> {
+/// Precomputed state provider and builder for block validation.
+///
+/// Stored as type-erased builder to avoid adding generics to `TreeCtx`.
+pub struct PrecomputedState {
```
another nit: calling this "state" is a bit confusing, it's not carrying state but instead the StateProvider and Builder. Maybe keep it obvious and go with "StateProviderAndBuilder" ?

yongkangc commented on crates/engine/tree/src/tree/payload_validator.rs:
```diff
@@ -60,13 +60,38 @@ pub struct TreeCtx<'a, N: NodePrimitives> {
     canonical_in_memory_state: &'a CanonicalInMemoryState<N>,
+    /// Optional precomputed state provider and builder to avoid redundant lookups.
+    /// This is set by [`crate::tree::EngineApiTreeHandler`] after validating parent state exists.
+    /// The builder is type-erased via `dyn Any` because adding a generic `P` to `TreeCtx` would
```
hmm yeah good point

yongkangc committed:
perf(engine): deduplicate state_provider_builder calls with type-safe TreeCtx

Avoid redundant state_provider_builder lookups during block validation by
passing both the prebuilt StateProviderBox and the StateProviderBuilder
through a type-safe TreeCtx.

Key changes:
- Add generic P to TreeCtx for type-safe builder storage (no downcast needed)
- Add Provider associated type to EngineValidator trait
- StateProviderBuilder::build() now consumes self, avoiding Vec clone
- TreeCtx::with_precomputed() builder pattern for cleaner API
- StateProviderAndBuilder holds both provider and builder

This eliminates duplicate blocks_by_hash traversals during payload validation
while maintaining full type safety.

Closes #20435

Amp-Thread-ID: https://ampcode.com/threads/T-019b2af5-3809-7349-958c-134380d47dfa

yongkangc committed:
chore: resolve merge conflict in payload_validator.rs

yongkangc committed:
chore: remove redundant comment

yongkangc committed:
bench(engine): add state_provider_builder micro-benchmark

yongkangc committed:
Revert "bench(engine): add state_provider_builder micro-benchmark"

This reverts commit 994ed1bbd8f77ebcbf0406f2a7e7f9100de8fa91.

yongkangc committed:
fix: remove Send + Sync bounds from Provider associated type

mattsse reviewed (commented):
these are quite a few changes just to remove a duplicate function
